### PR TITLE
Reduce Duplication Effort in zkpass-sdk Exploration

### PR DESCRIPTION
**Pull request by janice-laksana**
Asana : https://app.asana.com/0/1206389362636911/1207242505646832/f

This PR is comparing development branch with **[Automatic Script](https://github.com/gl-zkPass/zkpass-sdk/tree/f/automatic-script)** branch.
The **[git submodules](https://github.com/gl-zkPass/zkpass-sdk/tree/f/git-submodules)** and **[git subtree](https://github.com/gl-zkPass/zkpass-sdk/tree/f/git-subtree)** branch attached in the description below.

### Changes
1. add Automatic Script submodules.sh to copy related folder needed from zkpass repository into zkpass-sdk

### Issues
We need the SP1 source code in the zkpass-sdk repository, which is public. Therefore, we copied the SP1 source code from the zkpass repository, which is private, to the zkpass-sdk repository. We are not sure if this action is the best practice.

Other alternatives : **[git submodules](https://github.com/gl-zkPass/zkpass-sdk/tree/f/git-submodules)** and **[git subtree](https://github.com/gl-zkPass/zkpass-sdk/tree/f/git-subtree)**.

**Using Git Submodules** ([link](https://github.com/gl-zkPass/zkpass-sdk/tree/f/git-submodules))
Git submodules are used to include external repositories within a subdirectory of your own repository. However, submodules include the entire repository rather than specific directories. To include specific directories, you might need to use sparse checkout with submodules, but this is more complex and not always recommended for this use case.

With Git submodules, it adds all the folders from the remote repository. 
We can use sparse checkout to filter the folders we need from the remote repository, but the command is still experimental.

**Using Git Subtree** ([link](https://github.com/gl-zkPass/zkpass-sdk/tree/f/git-subtree))
The git subtree command doesn't directly support extracting only specific folders from a remote repository. It includes the entire repository history of the target subtree, but you can selectively use specific directories or files within your repository after pulling the subtree. However, if you only want to include specific directories without pulling the entire repository, you'd need to handle this with scripting or other tools.

Screenshot from 2024-05-16 12-57-00

**Summary**
- Automatic Script: Offers high customization and efficiency for copying specific files, but requires complex setup and maintenance. Best for highly specific use cases where only a few files are needed.
- Git Subtree: Provides a good balance between ease of use and granularity, preserving history and allowing integration of specific directories. Suitable for scenarios where only parts of a repository are needed.
- Git Submodules: Simplifies management of entire repositories as dependencies, preserving history and supporting nested submodules. Ideal for larger projects where entire repositories are needed.
- Git Submodules with Sparse Checkout: Offers selective checkout of submodule contents, combining the benefits of submodules with reduced disk space usage. However, sparse checkout is still experimental and may be unstable.

---

### Testing scenario
1. Automatic Scripting 
`. /rust/submodules.sh`

### Expected Result / Output
- No need to copy manually related folder needed from zkpass repository into zkpass-sdk

### CODE REVIEW

**Comment by khandar-william** [comment]
@janice-laksana masih "Draft"?

**Comment by janice-laksana** [comment]
Updated @khandar-william

File: rust/submodules.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+# Variables
+REPO_ZKPASS_URL="https://github.com/GDP-ADMIN/zkPass.git"
+TEMP_REPO_ZKPASS_DIR="tmp/zkpass"
```
Comment (khandar-william) [inline rust/submodules.sh:5]: bukan `/tmp/zkpass` ([/tmp directory](https://linuxhandbook.com/tmp-directory/))?

File: rust/submodules.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+DIRS_TO_COPY=("zkpass-client" "zkpass-demo/src" "zkpass-demo/tests" "zkpass-demo/Cargo.toml" "zkpass-core" "zkpass-query/types")  # The directories and files to copy
+
+# Clone repo A to a temporary directory
```
Comment (khandar-william) [inline rust/submodules.sh:8]:
```suggestion
# Clone zkPass repo to a temporary directory
```

File: rust/submodules.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+git clone "$REPO_ZKPASS_URL" "$TEMP_REPO_ZKPASS_DIR"
```
Comment (khandar-william) [inline rust/submodules.sh:9]: kalo dikasih error handling kayak gini jalan ga?
```sh
if git clone "$REPO_ZKPASS_URL" "$TEMP_REPO_ZKPASS_DIR"; then
  echo "Repository cloned successfully."
else
  echo "Failed to clone the repository."
  exit 1
fi
```

File: rust/submodules.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+# Variables
+REPO_ZKPASS_URL="https://github.com/GDP-ADMIN/zkPass.git"
```
Comment (khandar-william) [inline rust/submodules.sh:4]: zkpass-sdk adalah repo publik (seluruh dunia bisa lihat)
sedangkan `GDP-ADMIN/zkPass` ini adalah repo private (hanya gdp labs yg boleh lihat)
apakah boleh nulis URL repo private ini di repo publik?

File: rust/submodules.sh (hunk 1 above)
Comment (khandar-william) [reply rust/submodules.sh:5]: oh di akhir dihapus juga, kayaknya ga masalah

File: rust/submodules.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+TEMP_REPO_ZKPASS_DIR="tmp/zkpass"
+DIRS_TO_COPY=("zkpass-client" "zkpass-demo/src" "zkpass-demo/tests" "zkpass-demo/Cargo.toml" "zkpass-core" "zkpass-query/types")  # The directories and files to copy
+
```
Comment (khandar-william) [inline rust/submodules.sh:7]: tambah parameter branch yg di-copy dari repo sumber

**Comment by khandar-william** [comment]
> Automatic Scripting
> . /rust/submodules.sh

kalo gw jalanin script ini di folder `zkpass-sdk`, jadinya malah muncul directory `zkpass-sdk/zkpass-client` bukan `zkpass-sdk/rust/zkpass-client`

bisa diubah scriptnya untuk otomatis mendeteksi base directory adalah lokasi script itu sendiri?
jadi mau dijalanin dari `zkpass-sdk/` ataupun dari `zkpass-sdk/rust/` hasilnya tetap sama

File: rust/submodules.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    # If it's a file, copy the file
+    cp "$TEMP_REPO_ZKPASS_DIR/$DIR" "$DIR"
+  fi
+done
+
+# Clean up the temporary directory
+rm -rf "$TEMP_REPO_ZKPASS_DIR"
+rm -rf "tmp"  # Remove the tmp directory
+
+echo "Submodules have been successfully copied."
```
Comment (khandar-william) [inline rust/submodules.sh:46]: 1. saran nama file: `update-subdirectories.sh`
2. 
```suggestion
echo "Subdirectories have been successfully updated."
```

File: rust/.env.example [hunk 7]
```diff
@@ -0,0 +1,5 @@
+# Private repository URL
+REPO_ZKPASS_URL=https://github.com/GDP-ADMIN/zkPass.git
```
Comment (khandar-william) [inline rust/.env.example:2]: kaga ada gunanya kalo cuma dipisah .env begini
sama2 tetap masuk repository

bukan itu maksud gw, ini malah unecessary complexity

biarin tetap ada variable REPO_ZKPASS_URL di dalam script
tapi isinya string kosong atau `<put zkPass repo URL here>`

File: rust/submodules.sh (hunk 5 above)
Comment (khandar-william) [reply rust/submodules.sh]: ga perlu parameter source branch?

File: rust/.env.example [hunk 8]
```diff
@@ -0,0 +1,3 @@
+
```
Comment (khandar-william) [inline rust/.env.example:1]: balikin scriptnya sehingga ga perlu file .env sama sekali
kalo orang yg ngejalanin script itu perlu ubah sesuatu, langsung ubah variable di dalam scriptnya

File: rust/update-subdirectories.sh [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+if [ -f "$ENV_FILE" ]; then
+  source "$ENV_FILE"
+else
+  echo "Error: .env file not found!"
+  exit 1
+fi
+
+# Variables
+# REPO_ZKPASS_URL="<put zkPass repo URL here>"
+REPO_ZKPASS_URL="https://github.com/GDP-ADMIN/zkPass.git"
```
Comment (khandar-william) [inline rust/update-subdirectories.sh:15]: hanya begini aja
```suggestion
REPO_ZKPASS_URL="<put zkPass repo URL here>"
```

Comment (janice-laksana) [reply rust/update-subdirectories.sh:15]: oh yaa maap kepush.. lagi testing dlu mastiin dulu tp ke push..

File: rust/update-subdirectories.sh [hunk 10]
```diff
@@ -0,0 +1,55 @@
+#!/bin/bash
+
+# Variables
+REPO_ZKPASS_URL="<put zkPass repo URL here>"
+SCRIPT_DIR="$(cd "$(dirname "$0")" && pwd)"
+TEMP_REPO_ZKPASS_DIR="$SCRIPT_DIR/tmp/zkpass"
+TARGET_BASE_DIR="$SCRIPT_DIR"
+DIRS_TO_COPY="<put the directories and files to copy>"
```
Comment (khandar-william) [inline rust/update-subdirectories.sh:8]: untuk DIRS_TO_COPY langsung diisi gpp, karena nama directorynya memang sudah ada di repository ini

File: rust/update-subdirectories.sh [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
```
Comment (khandar-william) [inline rust/update-subdirectories.sh:9]: ga perlu ada variable untuk REPO_ZKPASS_BRANCH ?
apakah pasti selalu copy dari master?

**Pull request closed without merging**